File: frontend/lib/backend/document-intelligence.ts
const MAX_READ_BYTES = 2 * 1024 * 1024;

const NUMBER_PATTERNS: Record<string, RegExp[]> = {
  monthly_income: [
    /monthly[_\s-]*income\D*([0-9]+(?:\.[0-9]+)?)/i,
    /income\D*([0-9]+(?:\.[0-9]+)?)/i,
  ],
  monthly_expenses: [
    /monthly[_\s-]*expenses\D*([0-9]+(?:\.[0-9]+)?)/i,
    /expenses\D*([0-9]+(?:\.[0-9]+)?)/i,
  ],
  existing_debt: [
    /existing[_\s-]*debt\D*([0-9]+(?:\.[0-9]+)?)/i,
    /debt\D*([0-9]+(?:\.[0-9]+)?)/i,
  ],
  credit_utilization: [
    /credit[_\s-]*utilization\D*([0-9]+(?:\.[0-9]+)?)/i,
    /utilization\D*([0-9]+(?:\.[0-9]+)?)/i,
  ],
  delinquencies_12m: [
    /delinquencies[_\s-]*12m\D*([0-9]+)/i,
    /delinquencies\D*([0-9]+)/i,
  ],
  employment_months: [
    /employment[_\s-]*months\D*([0-9]+)/i,
    /employment\D*([0-9]+)\s*months/i,
  ],
  overdrafts_90d: [
    /overdrafts[_\s-]*90d\D*([0-9]+)/i,
    /overdrafts\D*([0-9]+)/i,
  ],
};

export interface UploadedFileInput {
  filename: string;
  buffer: Buffer;
}

export interface DocumentAnalysis {
  evidence_refs: string[];
  document_text: string;
  document_signals: {
    numeric_hints: Record<string, number>;
    consistency_flags: string[];
    coverage_score: number;
    income_verified_from_docs: boolean;
    unreadable_files: string[];
    stored_paths?: string[];
  };
}

function safeDecode(raw: Buffer): string {
  return raw.toString("utf-8");
}

function extractFromCsv(text: string): Record<string, number> {
  const out: Record<string, number> = {};
  const lines = text.trim().split("\n");
  if (lines.length < 2) return out;

  const headers = lines[0].split(",").map((h) => h.trim().toLowerCase());
  const values = lines[1].split(",").map((v) => v.trim());

  for (const key of Object.keys(NUMBER_PATTERNS)) {
    const idx = headers.findIndex((h) => h.includes(key.replace(/_/g, "")) || h === key);
    if (idx >= 0 && values[idx]) {
      const num = parseFloat(values[idx]);
      if (!isNaN(num)) out[key] = num;
    }
    if (headers.includes(key) && values[headers.indexOf(key)]) {
      const num = parseFloat(values[headers.indexOf(key)]);
      if (!isNaN(num)) out[key] = num;
    }
  }

  return out;
}

function extractFromJson(text: string): Record<string, number> {
  const out: Record<string, number> = {};
  try {
    const payload = JSON.parse(text);
    const source = Array.isArray(payload) ? payload[0] : payload;
    if (!source || typeof source !== "object") return out;

    for (const key of Object.keys(NUMBER_PATTERNS)) {
      if (key in source) {
        const num = parseFloat(String(source[key]));
        if (!isNaN(num)) out[key] = num;
      }
    }
  } catch {
    // ignore
  }
  return out;
}

function extractWithRegex(text: string): Record<string, number> {
  const out: Record<string, number> = {};
  for (const [key, patterns] of Object.entries(NUMBER_PATTERNS)) {
    for (const pattern of patterns) {
      const match = text.match(pattern);
      if (match?.[1]) {
        const num = parseFloat(match[1]);
        if (!isNaN(num)) {
          out[key] = num;
          break;
        }
      }
    }
  }
  return out;
}

function normalizeHints(hints: Record<string, number>): Record<string, number> {
  const normalized = { ...hints };
  if (normalized.credit_utilization !== undefined) {
    const util = normalized.credit_utilization;
    normalized.credit_utilization = util > 1 ? util / 100 : util;
  }
  for (const intKey of [
    "delinquencies_12m",
    "employment_months",
    "overdrafts_90d",
  ] as const) {
    if (normalized[intKey] !== undefined) {
      normalized[intKey] = Math.round(normalized[intKey]);
    }
  }
  return normalized;
}

function signalsFromText(text: string) {
  const lowered = text.toLowerCase();
  const incomeVerified =
    lowered.includes("income verified") || lowered.includes("verified income");
  const consistencyFlags: string[] = [];

  if (lowered.includes("fraud") || lowered.includes("forg")) {
    consistencyFlags.push("POTENTIAL_FRAUD_SIGNAL");
  }
  if (lowered.includes("inconsistent") || lowered.includes("mismatch")) {
    consistencyFlags.push("DOCUMENT_DATA_MISMATCH");
  }

  return {
    income_verified_from_docs: incomeVerified,
    consistency_flags: consistencyFlags,
    numeric_hints: extractWithRegex(text),
  };
}

export function analyzeUploadedDocuments(
  files: UploadedFileInput[]
): DocumentAnalysis {
  const evidenceRefs: string[] = [];
  const allTextParts: string[] = [];
  const mergedHints: Record<string, number> = {};
  const consistencyFlags: string[] = [];
  const unreadableFiles: string[] = [];

  for (const file of files) {
    const filename = file.filename || "unnamed_document";
    evidenceRefs.push(filename);

    const raw = file.buffer.subarray(0, MAX_READ_BYTES);
    const ext = filename.split(".").pop()?.toLowerCase() ?? "";

    if (!raw.length) {
      unreadableFiles.push(filename);
      continue;
    }

    const text = safeDecode(raw);
    if (!text.trim() && [".pdf", ".png", ".jpg", ".jpeg"].some((e) => ext === e.slice(1))) {
      unreadableFiles.push(filename);
      continue;
    }

    allTextParts.push(text);

    let perFileHints: Record<string, number> = {};
    if (ext === "json") perFileHints = extractFromJson(text);
    else if (ext === "csv") perFileHints = extractFromCsv(text);

    const signal = signalsFromText(text);
    perFileHints = { ...perFileHints, ...signal.numeric_hints };

    Object.assign(mergedHints, perFileHints);
    for (const flag of signal.consistency_flags) {
      if (!consistencyFlags.includes(flag)) consistencyFlags.push(flag);
    }
  }

  const normalizedHints = normalizeHints(mergedHints);
  const joinedText = allTextParts.join("\n");
  const coverageScore = Math.min(1, Object.keys(normalizedHints).length / 7);
  const loweredAll = joinedText.toLowerCase();
  const incomeVerifiedFromDocs =
    loweredAll.includes("income verified") ||
    loweredAll.includes("verified income");

  return {
    evidence_refs: evidenceRefs,
    document_text: joinedText.slice(0, 12000),
    document_signals: {
      numeric_hints: normalizedHints,
      consistency_flags: consistencyFlags,
      coverage_score: Math.round(coverageScore * 100) / 100,
      income_verified_from_docs: incomeVerifiedFromDocs,
      unreadable_files: unreadableFiles,
    },
  };
}

export { inferDocumentType } from "@/lib/utils";

File: backend/app/services/document_intelligence.py
from __future__ import annotations

import csv
import io
import json
import re
from pathlib import Path
from uuid import uuid4
from typing import Any

from fastapi import UploadFile

# Hard cap to protect memory for hackathon demo environment.
MAX_READ_BYTES = 2 * 1024 * 1024
UPLOADS_DIR = Path(__file__).resolve().parents[2] / "uploads"


NUMBER_PATTERNS = {
    "monthly_income": [r"monthly[_\s-]*income\D*([0-9]+(?:\.[0-9]+)?)", r"income\D*([0-9]+(?:\.[0-9]+)?)"],
    "monthly_expenses": [r"monthly[_\s-]*expenses\D*([0-9]+(?:\.[0-9]+)?)", r"expenses\D*([0-9]+(?:\.[0-9]+)?)"],
    "existing_debt": [r"existing[_\s-]*debt\D*([0-9]+(?:\.[0-9]+)?)", r"debt\D*([0-9]+(?:\.[0-9]+)?)"],
    "credit_utilization": [r"credit[_\s-]*utilization\D*([0-9]+(?:\.[0-9]+)?)", r"utilization\D*([0-9]+(?:\.[0-9]+)?)"],
    "delinquencies_12m": [r"delinquencies[_\s-]*12m\D*([0-9]+)", r"delinquencies\D*([0-9]+)"],
    "employment_months": [r"employment[_\s-]*months\D*([0-9]+)", r"employment\D*([0-9]+)\s*months"],
    "overdrafts_90d": [r"overdrafts[_\s-]*90d\D*([0-9]+)", r"overdrafts\D*([0-9]+)"],
}


def _safe_decode(raw: bytes) -> str:
    return raw.decode("utf-8", errors="ignore")


def _extract_from_csv(text: str) -> dict[str, float]:
    out: dict[str, float] = {}
    reader = csv.DictReader(io.StringIO(text))
    for row in reader:
        for key in NUMBER_PATTERNS:
            if key in row and row[key] not in (None, ""):
                try:
                    out[key] = float(row[key])
                except ValueError:
                    continue
        if out:
            break
    return out


def _extract_from_json(text: str) -> dict[str, float]:
    out: dict[str, float] = {}
    try:
        payload = json.loads(text)
    except json.JSONDecodeError:
        return out

    if isinstance(payload, dict):
        source = payload
    elif isinstance(payload, list) and payload and isinstance(payload[0], dict):
        source = payload[0]
    else:
        return out

    for key in NUMBER_PATTERNS:
        if key in source:
            try:
                out[key] = float(source[key])
            except (TypeError, ValueError):
                continue
    return out


def _extract_with_regex(text: str) -> dict[str, float]:
    out: dict[str, float] = {}
    lowered = text.lower()
    for key, patterns in NUMBER_PATTERNS.items():
        for pattern in patterns:
            match = re.search(pattern, lowered)
            if match:
                try:
                    out[key] = float(match.group(1))
                    break
                except ValueError:
                    continue
    return out


def _normalize_hints(hints: dict[str, float]) -> dict[str, Any]:
    normalized: dict[str, Any] = dict(hints)
    if "credit_utilization" in normalized:
        util = float(normalized["credit_utilization"])
        # Accept either ratio form (0.62) or percentage form (62).
        normalized["credit_utilization"] = util / 100 if util > 1 else util

    for int_key in ("delinquencies_12m", "employment_months", "overdrafts_90d"):
        if int_key in normalized:
            normalized[int_key] = int(round(float(normalized[int_key])))
    return normalized


def _signals_from_text(text: str) -> dict[str, Any]:
    lowered = text.lower()
    income_verified = "income verified" in lowered or "verified income" in lowered
    consistency_flags: list[str] = []

    if "fraud" in lowered or "forg" in lowered:
        consistency_flags.append("POTENTIAL_FRAUD_SIGNAL")
    if "inconsistent" in lowered or "mismatch" in lowered:
        consistency_flags.append("DOCUMENT_DATA_MISMATCH")

    hints = _extract_with_regex(text)
    return {
        "income_verified_from_docs": income_verified,
        "consistency_flags": consistency_flags,
        "numeric_hints": hints,
    }


async def analyze_uploaded_documents(files: list[UploadFile]) -> dict[str, Any]:
    evidence_refs: list[str] = []
    all_text_parts: list[str] = []
    merged_hints: dict[str, float] = {}
    consistency_flags: list[str] = []
    unreadable_files: list[str] = []

    for upload in files:
        evidence_refs.append(upload.filename or "unnamed_document")
        raw = await upload.read(MAX_READ_BYTES)
        await upload.seek(0)
        ext = Path(upload.filename or "").suffix.lower()

        if not raw:
            unreadable_files.append(upload.filename or "unnamed_document")
            continue

        text = _safe_decode(raw)
        if not text.strip() and ext in {".pdf", ".png", ".jpg", ".jpeg"}:
            unreadable_files.append(upload.filename or "unnamed_document")
            continue

        all_text_parts.append(text)

        per_file_hints: dict[str, float] = {}
        if ext == ".json":
            per_file_hints.update(_extract_from_json(text))
        elif ext == ".csv":
            per_file_hints.update(_extract_from_csv(text))

        signal = _signals_from_text(text)
        per_file_hints.update(signal["numeric_hints"])

        for key, value in per_file_hints.items():
            merged_hints[key] = value

        for flag in signal["consistency_flags"]:
            if flag not in consistency_flags:
                consistency_flags.append(flag)

    merged_hints = _normalize_hints(merged_hints)
    joined_text = "\n".join(all_text_parts)
    coverage_score = min(1.0, len(merged_hints) / 7)

    lowered_all = joined_text.lower()
    income_verified_from_docs = "income verified" in lowered_all or "verified income" in lowered_all

    return {
        "evidence_refs": evidence_refs,
        "document_text": joined_text[:12000],
        "document_signals": {
            "numeric_hints": merged_hints,
            "consistency_flags": consistency_flags,
            "coverage_score": round(coverage_score, 2),
            "income_verified_from_docs": income_verified_from_docs,
            "unreadable_files": unreadable_files,
        },
    }


async def persist_uploaded_documents(files: list[UploadFile], case_id: str) -> list[str]:
    case_dir = UPLOADS_DIR / case_id
    case_dir.mkdir(parents=True, exist_ok=True)
    stored_paths: list[str] = []

    for upload in files:
        original_name = upload.filename or "unnamed_document"
        safe_name = f"{uuid4().hex[:8]}_{Path(original_name).name}"
        target_path = case_dir / safe_name

        raw = await upload.read(MAX_READ_BYTES)
        await upload.seek(0)
        target_path.write_bytes(raw)
        stored_paths.append(str(target_path))

    return stored_paths
